### PR TITLE
Gnc

### PR DESCRIPTION
**Pull request by lucacarlone**
Added Graduated Non-Convexity (GNC) as a nonlinear optimizer in GTSAM. This is an implementation of the GNC paper (https://arxiv.org/pdf/1909.08605.pdf), but it uses iterative solvers in the inner loop instead of non-minimal global solvers.

Outstanding tasks:
- [x] reintroduce “using BaseOptimizer = BaseOptimizerParameters::OptimizerType;”  to let the user choose the iterative solver for the inner loop.
- [x] move GncOptimizer and its params to a proper class/file when ready.
- [x] populate the TLS option (now the code only implements GM = Geman McClure).
- [x] double-check that the 0.5 in front of the factors in gtsam does not require a modification to the weight update rule.

### CODE REVIEW

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:612]: Just say `GncParams<GaussNewtonParams> gncParams;` ?

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:39]: Probably need to move to a .h file now

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:73]: add comments to all methods, at least `/// one-liner doxygen`

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:53]: Oh ?

**Review by dellaert — COMMENTED** [review]
I left some comments. What is the main templating issue you wanted me to look at ?

File: tests/testGncOptimizer.cpp
Comment (lucacarlone) [reply tests/testGncOptimizer.cpp:612]: solved

File: tests/testGncOptimizer.cpp
Comment (lucacarlone) [reply tests/testGncOptimizer.cpp:53]: yes, this exactly the problem we need to solve. I can do:
using BaseOptimizer = BaseOptimizerParameters::OptimizerType;
Then I can go to LevenbergMarquardtParams and define OptimizerType as LevenbergMarquardtOptimizer.
However, LevenbergMarquardtParams does not know about LevenbergMarquardtOptimizer, so I'm not sure how to make that define work. did you have something different in mind?

Comment (dellaert) [reply tests/testGncOptimizer.cpp]: I think you need to modify LevenbergMarquardtParams to add the typedef there

Comment (lucacarlone) [reply tests/testGncOptimizer.cpp]: ok I'll try that, but I suspect an easier way would be to template on the Optimizer instead of the Params.

File: gtsam/nonlinear/GncOptimizer.h
Comment (lucacarlone) [inline gtsam/nonlinear/GncOptimizer.h:270]: @jingnanshi : I don't think mu can be negative. does it make sense to have a lower bound for mu in this and the GM case? it the max residual is below barcsq, it means there is no outlier and maybe we can just "return"

Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h:270]: @lucacarlone I completely agree. I put the check outside the initializeMu function (see line 210 to 217) so that we can easily just return the current result.

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h:270]: great, @jingnanshi !

File: gtsam/nonlinear/GncOptimizer.h
Comment (lucacarlone) [inline gtsam/nonlinear/GncOptimizer.h:297]: @jingnanshi : this does not make much sense, since the decrease in mu is constant. maybe @pantonante has a better check?

**Comment by lucacarlone** [comment]
@jingnanshi @dellaert @pantonante : I fixed the outstanding issue with optimizerType. I also added a couple of unit tests (and comments) to make sure the inlier threshold is set correctly. @jingnanshi : the only thing I'm not convinced about is the stopping condition for TLS - please check my comment. Besides that, this PR is ready for review!

File: gtsam/nonlinear/GncOptimizer.h
Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h:304]: @lucacarlone : good catch! this should be cost - cost_prev

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h:304]: I think Pasquale checks if the weights are (close to) binary, which is the case after TLS converges.

Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h:304]: @lucacarlone : I just checked Pasquale's email, and he suggested using the relative change of the weighted sum of the squared residuals, which is what I used in TEASER++. I think I'll implement a checkConvergence function that checks mu or relative cost change depending on loss type.

Comment (pantonante) [reply gtsam/nonlinear/GncOptimizer.h:304]: Yes, I suggest to use the relative change of the weighted sum of the squared residuals because it works well in practice. Here seems you are checking that they are not changing much, this doesn't work for TLS.

File: gtsam/nonlinear/GncOptimizer.h
Comment (pantonante) [inline gtsam/nonlinear/GncOptimizer.h:65]: Probably we want TLS as default loss type.

File: gtsam/nonlinear/GncOptimizer.h
Comment (pantonante) [inline gtsam/nonlinear/GncOptimizer.h:345]: This will work well, but it looks weird that we use two very different stopping criteria for the two robust costs. I would use the cost based convergence test in both cases as default and add a parameter to support three convergence tests:
1. cost based only (default)
2. weighs based only
3. [optional] cost based OR weights base (whatever converges first)
I think this would add clarity overall.

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h:345]: for simplicity, I would add both stopping conditions (for both costs) and give the user control of the tolerances for stopping, such that s/he can enable/disable the conditions s/he prefers.

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h:345]: @jingnanshi : let me know if you can take a look at this, otherwise I can do it later in December.

Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h:345]: @lucacarlone @pantonante For the TLS case, since mu is increasing, it is difficult to check for mu (unless I have misunderstood it). What do you guys recommend?

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h:345]: @jingnanshi : Pasquale was suggesting two checks I think:
1) relative change in cost: e.g., | cost_t - cost_t-1 | /  |cost_t | < params_.relativeCostTol (then stop)
2) weights: e.g., if | weight - round(weights) | < params_.weightConvergenceTol  (then stop)
In the latter round(weights) rounds the entries of weights in {0;1} and this is mostly effective for TLS (it checks whether you have a clear decision inlier/outliers).

Note: you can also keep the stopping condition based on mu for GM, while there is no such condition for TLS

File: gtsam/nonlinear/GncOptimizer.h
Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h]: @jingnanshi : I would implement this in the main loop (before "if (checkMuConvergence(mu, mu_prev))"): I would simply add "if (checkCostConvergence(..)  OR checkWeightsConvergence(..))" and implement these in a separate function. Let me know if you want me to take a stab - I can try to do that tonight

Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h]: @lucacarlone : That makes sense. I won't be able to do it tonight unfortunately so feel free to take a look if you have time!

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h]: @jingnanshi @pantonante : I fixed the stopping conditions and a few other things (including setting TLS to default) - please review carefully and let me know if you have more comments.

Comment (jingnanshi) [reply gtsam/nonlinear/GncOptimizer.h]: @lucacarlone : Had two very minor fixes. Everything looks good!

Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h]: perfect - thanks @jingnanshi !

File: gtsam/nonlinear/GaussNewtonOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GaussNewtonOptimizer.h:32]: Change to `using OptimizerType = GaussNewtonOptimizer`

File: gtsam/nonlinear/LevenbergMarquardtParams.h
Comment (dellaert) [inline gtsam/nonlinear/LevenbergMarquardtParams.h:45]: Change to `using OptimizerType = LevenbergMarquardtOptimizer `

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:47]: Do we need the <> here (and below) or can the type be derived from the argument? I think it can and should be omitted.

File: tests/testGncOptimizer.cpp
Comment (dellaert) [inline tests/testGncOptimizer.cpp:84]: If default. I think we should test here the use of 
```
GncParams<LevenbergMarquardtParams> gncParams;
```
i.e. without arguments

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:36]: There is enough code here to warrant a separate file for GncParams

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:43]: GNC here is redundant

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:48]: I propose to make this its own class called `GncLossType`, not an inner class. Will make things less verbose.

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:63]: Single line comments after a line should use doxygen `///< Optimization parameters...
Also, it seems this line is too long so probably you did not format with Google style? Please do so, and if comment is not single line, add it above with doxygen /** ... */

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:47]: Single line doxygen should use `///`

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:65]: `///< Default loss.`

Note capitalization and punctuation :-)

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:160]: Document

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:195]: Here and below: add return. Make sure to format with clang-format or cpp-format w Google style.

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [inline gtsam/nonlinear/GncOptimizer.h:305]: Capitalize and add period to all comments.

**Review by dellaert — COMMENTED** [review]
Awesome! First round, will look again after comments resolved :-)

File: tests/testGncOptimizer.cpp
Comment (lucacarlone) [reply tests/testGncOptimizer.cpp:47]: @dellaert : it does not work if I omit - should I do something else to make it derive the type?

File: gtsam/nonlinear/GncOptimizer.h
Comment (lucacarlone) [reply gtsam/nonlinear/GncOptimizer.h]: @dellaert : what do you envision we should put in the new class? this is currently just an enum

File: tests/testGncOptimizer.cpp
Comment (dellaert) [reply tests/testGncOptimizer.cpp:47]: Hmmm, I had expected that to work, but maybe it's only since C++17: https://en.cppreference.com/w/cpp/language/class_template_argument_deduction

File: gtsam/nonlinear/GncOptimizer.h
Comment (dellaert) [reply gtsam/nonlinear/GncOptimizer.h]: Just before the params class: otherwise you have to append the entire (templated) parameter class name which is unwieldy.

File: gtsam/nonlinear/GncParams.h
Comment (dellaert) [inline gtsam/nonlinear/GncParams.h:80]: Add newline between methods...

**Review by dellaert — APPROVED** [review]
Very cool! 
Just a few more code nits, feel free to merge after addressed ;-)

Now we need to apply it to Shonan and compare with robuts rotation averaging, right @jingwuOUO  :-) ?

**Comment by lucacarlone** [comment]
all set! I'll merge after @pantonante and @jingnanshi  approve!

**Review by jingnanshi — APPROVED** [review]
:+1:

**Review by pantonante — APPROVED** [review]

**Pull request merged**
